x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/ChunkedTrainedModelRestorer.java

```diff
@@ -0,0 +1,222 @@
+     * there is no error checking for duplicate or missing docs the consumer should handle
+     * those errors.
+     *
+     * Depending on the search size multiple searches may be made.
+     *
+     * @param modelConsumer    Consumes model definition docs
+     * @param successConsumer  Called when all docs have been returned or the loading is cancelled
+     * @param errorConsumer    In the event of an error
+     */
+    public void restoreModelDefinition(CheckedFunction<TrainedModelDefinitionDoc, Boolean, IOException> modelConsumer,
+                                       Consumer<Boolean> successConsumer,
+                                       Consumer<Exception> errorConsumer) {
+
+        SearchRequest searchRequest = buildSearch(client, modelId, index, searchSize);
+
+        executorService.execute(() -> doSearch(searchRequest, modelConsumer, successConsumer, errorConsumer));
+    }
+
+    private void doSearch(SearchRequest searchRequest,
+                          CheckedFunction<TrainedModelDefinitionDoc, Boolean, IOException> modelConsumer,
+                          Consumer<Boolean> successConsumer,
+                          Consumer<Exception> errorConsumer) {
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, ActionListener.wrap(
+            searchResponse -> {
+                if (searchResponse.getHits().getHits().length == 0) {
+                    errorConsumer.accept(new ResourceNotFoundException(
+                        Messages.getMessage(Messages.MODEL_DEFINITION_NOT_FOUND, modelId)));
+                    return;
+                }
+
+                // Set lastNum to a non-zero to prevent an infinite loop
+                // search after requests in the absolute worse case where
+                // it has all gone wrong.
+                // Docs are numbered 0..N. we must have seen at least
+                // this many docs so far.
+                int lastNum = numDocsWritten -1;
+                for (SearchHit hit : searchResponse.getHits().getHits()) {
+                    try {
+                        TrainedModelDefinitionDoc doc =
+                            parseModelDefinitionDocLenientlyFromSource(hit.getSourceRef(), modelId, xContentRegistry);
+                        lastNum = doc.getDocNum();
+
+                        boolean continueSearching = modelConsumer.apply(doc);
+                        if (continueSearching == false) {
+                            // signal we are finished
+                            successConsumer.accept(Boolean.FALSE);
+                            return;
+                        }
+
+                    } catch (IOException e) {
+                        logger.error(new ParameterizedMessage("[{}] error writing model definition", modelId), e);
+                        errorConsumer.accept(e);
+                        return;
+                    }
+                }
```

Comment:
It would be good to make sure that these docs don't spread across numerous indices. Or, we make sure that they are in the correct order. 

Given the query, if these docs spread numerous indices, we don't have guarantees that the docs are in the right order.

Reply:
Yeah I've shown that in the test here 

https://github.com/elastic/elasticsearch/blob/7904556a1f9669e1f523768646fabeb33f618c2b/x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/ChunkedTrainedModelRestorerIT.java#L162

Reply:
This test doesn't cover the failure scenario:

`doc 0 in index 2`
`doc 0 in index 1`
`doc 1 in index 2`

This may not be possible given  the current APIs, but that is the scenario which I was concerned with.

Reply:
Given that the query filters on `model_id` how could there be 2 `doc 0`s for a given model? Isn't there a check against creating 2 models with the same name?

Reply:
yes, there are checks everywhere. This is simply another one.